### PR TITLE
Sharing

### PR DESCRIPTION
**Pull request by tap2k**
Added UserAuthorityList and UserAuthorityInstance API endpoints along with associated tests. I am not testing object visibility until we implement it in the model.

### CODE REVIEW

File: deploy_tools/settings_local.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 ADMIN_EMAILS = ['your_email@gmail.com',]
 
 # Custom Local Variables
-SERVER_HOST = 'localhost:7777'
+SERVER_HOST = 'daya.ischool.berkeley.edu:7777'
```
Comment (vanwars) [inline deploy_tools/settings_local.py]: Reverting this. Make sure you change the settings_local.py in apps directory, not the deploy_tools one.

File: deploy_tools/settings_local.py [hunk 2]
```diff
@@ -61,3 +61,5 @@
 INTERNAL_IPS = ('127.0.0.1', '10.0.2.2') #note the 10.0.2.2 is the IP for Vagrant connections
 MIDDLEWARE_CLASSES += ('debug_toolbar.middleware.DebugToolbarMiddleware',)
 '''
+
```
Comment (vanwars) [inline deploy_tools/settings_local.py]: No sendgrid API key in source code, please. We might want to regenerate this.

File: apps/site/models/permissions.py [hunk 3]
```diff
@@ -134,6 +134,16 @@ def to_dict(self):
     def __unicode__(self):
         return self.user.username
 
+    # TODO: This seems a little dicey if Im not inheriting from BasePermissions?
```
Comment (vanwars) [inline apps/site/models/permissions.py]: I think this is good -- it's relying on the project's can_view/edit/manage methods, which do rely on BasePermissions.

**Comment by vanwars** [comment]
Hi Tap,

I did some refactoring, based on your original plan, and changed the sharing URLs to be:
/api/0/projects/<project_id>/users/<user_id>/ (just as you originally wanted it), so that it's consistent with similar patterns (e.g. /api/0/markers/<marker_id>/photos/<photo_id>/. I also added quite a few tests for permissions checks.

I'm happy with the changes now, but it would be great if you could review them also (particularly the new permissions rules) and see if they're enforcing the rules we want. Thanks!

**Comment by tap2k** [comment]
Some questions / comments:
- In apps/site/api/views/sharing_views.py line number 56, we never access that project object so do we need it?
- In apps/site/models/permissions.py, what is the purpose of the `self.user == user` checks on lines 140, 145, 148?
- In apps/site/api/permissions.py line number 18, shouldnt that be can_manage instead of can_view?

**Comment by tap2k** [comment]
Im also getting a couple of testing errors w/ the celery code. Am I missing some config? I did a fresh vagrant install:

ERROR: test_update_print_using_put (localground.apps.site.api.tests.mapimage_tests.ApiMapImageDetailTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/localground/apps/site/api/tests/mapimage_tests.py", line 109, in test_update_print_using_put
    content_type="application/json"
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 546, in put
    secure=secure, **extra)
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 339, in put
    secure=secure, **extra)
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 379, in generic
    return self.request(**r)
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 466, in request
    six.reraise(*exc_info)
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 452, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 449, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/generics.py", line 287, in put
    return self.update(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py", line 69, in update
    self.perform_update(serializer)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py", line 73, in perform_update
    serializer.save()
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/serializers.py", line 160, in save
    self.instance = self.update(self.instance, validated_data)
  File "/localground/apps/site/api/serializers/mapimage_serializer.py", line 164, in update
    result = process_map.delay(self.instance)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 453, in delay
    return self.apply_async(args, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 565, in apply_async
    **dict(self._get_exec_options(), **options)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/base.py", line 354, in send_task
    reply_to=reply_to or self.oid, **options
  File "/usr/local/lib/python2.7/dist-packages/celery/app/amqp.py", line 310, in publish_task
    **kwargs
  File "/usr/local/lib/python2.7/dist-packages/kombu/messaging.py", line 172, in publish
    routing_key, mandatory, immediate, exchange, declare)
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 457, in _ensured
    interval_max)
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 369, in ensure_connection
    interval_start, interval_step, interval_max, callback)
  File "/usr/local/lib/python2.7/dist-packages/kombu/utils/__init__.py", line 246, in retry_over_time
    return fun(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 237, in connect
    return self.connection
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 742, in connection
    self._connection = self._establish_connection()
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 697, in _establish_connection
    conn = self.transport.establish_connection()
  File "/usr/local/lib/python2.7/dist-packages/kombu/transport/pyamqp.py", line 116, in establish_connection
    conn = self.Connection(**opts)
  File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 165, in __init__
    self.transport = self.Transport(host, connect_timeout, ssl)
  File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 186, in Transport
    return create_transport(host, connect_timeout, ssl)
  File "/usr/local/lib/python2.7/dist-packages/amqp/transport.py", line 299, in create_transport
    return TCPTransport(host, connect_timeout)
  File "/usr/local/lib/python2.7/dist-packages/amqp/transport.py", line 95, in __init__
    raise socket.error(last_err)
error: [Errno 111] Connection refused

======================================================================
ERROR: test_create_mapimage_using_post (localground.apps.site.api.tests.mapimage_tests.ApiMapImageListTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/localground/apps/site/api/tests/mapimage_tests.py", line 52, in test_create_mapimage_using_post
    HTTP_X_CSRFTOKEN=self.csrf_token
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 512, in post
    secure=secure, **extra)
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 313, in post
    secure=secure, **extra)
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 379, in generic
    return self.request(**r)
  File "/usr/local/lib/python2.7/dist-packages/django/test/client.py", line 466, in request
    six.reraise(*exc_info)
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 452, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 449, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/generics.py", line 244, in post
    return self.create(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py", line 20, in create
    self.perform_create(serializer)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/mixins.py", line 25, in perform_create
    serializer.save()
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/serializers.py", line 165, in save
    self.instance = self.create(validated_data)
  File "/localground/apps/site/api/serializers/mapimage_serializer.py", line 107, in create
    result = process_map.delay(self.instance)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 453, in delay
    return self.apply_async(args, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 565, in apply_async
    **dict(self._get_exec_options(), **options)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/base.py", line 354, in send_task
    reply_to=reply_to or self.oid, **options
  File "/usr/local/lib/python2.7/dist-packages/celery/app/amqp.py", line 310, in publish_task
    **kwargs
  File "/usr/local/lib/python2.7/dist-packages/kombu/messaging.py", line 172, in publish
    routing_key, mandatory, immediate, exchange, declare)
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 457, in _ensured
    interval_max)
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 369, in ensure_connection
    interval_start, interval_step, interval_max, callback)
  File "/usr/local/lib/python2.7/dist-packages/kombu/utils/__init__.py", line 246, in retry_over_time
    return fun(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 237, in connect
    return self.connection
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 742, in connection
    self._connection = self._establish_connection()
  File "/usr/local/lib/python2.7/dist-packages/kombu/connection.py", line 697, in _establish_connection
    conn = self.transport.establish_connection()
  File "/usr/local/lib/python2.7/dist-packages/kombu/transport/pyamqp.py", line 116, in establish_connection
    conn = self.Connection(**opts)
  File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 165, in __init__
    self.transport = self.Transport(host, connect_timeout, ssl)
  File "/usr/local/lib/python2.7/dist-packages/amqp/connection.py", line 186, in Transport
    return create_transport(host, connect_timeout, ssl)
  File "/usr/local/lib/python2.7/dist-packages/amqp/transport.py", line 299, in create_transport
    return TCPTransport(host, connect_timeout)
  File "/usr/local/lib/python2.7/dist-packages/amqp/transport.py", line 95, in __init__
    raise socket.error(last_err)
error: [Errno 111] Connection refused

----------------------------------------------------------------------
Ran 350 tests in 58.666s

FAILED (errors=2)

**Comment by vanwars** [comment]
Re: sharing questions...I will take a look shortly.

Re: celery errors, please make sure the following packages are installed:

sudo apt-get -y install redis-server rabbitmq-server

# we use supervisor to run our celery worker
sudo apt-get -y install supervisor

# install python packages:
sudo pip install django-celery==3.1.17
sudo pip install flower==0.9.0

# move config files:
sudo cp /localground/deploy_tools/celeryd.conf
/etc/supervisor/conf.d/celeryd.conf
sudo mkdir /var/log/celery
sudo cp /localground/deploy_tools/flower.conf
/etc/supervisor/conf.d/flower.conf

# restart services
sudo service supervisor restart
sudo supervisorctl restart celery

**Comment by tap2k** [comment]
OK, but shouldnt these have been installed by the vagrant script?

**Comment by vanwars** [comment]
Yes. They're in the deploy script, but your most recent pull included a
merge w/master (which has some of the new celery stuff w/Flower). If you
didn't re-provision your vagrant, then the newer package installs wouldn't
have run. Know what I mean?

**Comment by tap2k** [comment]
I did re-provision. I always test in a completely fresh tree and vagrant box. Maybe its because I have local changes in apps/settings_local.py?

**Comment by vanwars** [comment]
Did you just re-provision with the last pull you did?

**Comment by tap2k** [comment]
Yes. I can try it again. I was missing the SENDGRID key - could that be it?

btw installing those packages fixed the errors.

**Comment by vanwars** [comment]
In response to your Qs:
Some questions / comments:
> In apps/site/api/views/sharing_views.py line number 56, we never access that project object so do we need it?

No. It's gone.

> In apps/site/models/permissions.py, what is the purpose of the self.user == user checks on lines 140, 145, 148?

I figured users should be able to remove themselves from projects (delete), view their own permissions status on a project (can_view), and assign themselves a lower permissions status (can_edit).

> In apps/site/api/permissions.py line number 18, shouldn't that be can_manage instead of can_view?

That just determines whether a user can view one's own permission object (which I think they should be able to). There are additional checks in sharing_views.py that ensure edit / delete permissions.

> Should we add a test that checks whether a user with only view permissions cannot share the project using using the list or detail views?

Yes.

**Pull request merged**